packages/puppeteer-core/src/api/Locator.ts

```diff
@@ -0,0 +1,374 @@
+/**
+ * Copyright 2023 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AbortError, TimeoutError} from '../common/Errors.js';
+import {EventEmitter} from '../common/EventEmitter.js';
+import type {MouseButton} from '../common/Input.js';
+import {debugError} from '../common/util.js';
+import {isErrorLike} from '../util/ErrorLike.js';
+
+import {ElementHandle, BoundingBox} from './ElementHandle.js';
+import type {Page} from './Page.js';
+
+/**
+ * @internal
+ */
+export interface LocatorOptions {
+  /**
+   * Whether to wait for the element to be `visible` or `hidden`.
+   */
+  visibility: 'hidden' | 'visible';
+  /**
+   * Total timeout for the entire locator operation.
+   */
+  timeout: number;
+  /**
+   * Timeout for individual operations inside the locator. On errors the
+   * operation is retried as long as {@link LocatorOptions.timeout} is not
+   * exceeded. This timeout should be generally much lower as locating an
+   * element means multiple asynchronious operations.
+   */
+  operationTimeout: number;
+}
+
+/**
+ * @internal
+ */
+type ActionCondition = (
+  element: ElementHandle,
+  signal: AbortSignal
+) => Promise<void>;
+
+/**
+ * @internal
+ */
+export interface ActionOptions {
+  signal?: AbortSignal;
+  conditions: ActionCondition[];
+}
+
+/**
+ * All the events that a locator instance may emit.
+ *
+ * @internal
+ */
+export enum LocatorEmittedEvents {
```

Comment:
```suggestion
export enum LocatorEmittedEvent {
```

Reply:
that suggestion does not follow our existing conventions.